frontend/src/lib/lemon-ui/LemonSelectMultiple/LemonSelectMultiple.tsx

```diff
@@ -1,121 +1,264 @@
+    disableFiltering?: boolean
+    mode?: 'single' | 'multiple'
+    allowCustomValues?: boolean
+    onChange?: (newValue: string[]) => void
+    onInputChange?: (newValue: string) => void
     'data-attr'?: string
 }
 
-export type LabelInValue = { value: string; label: ReactNode }
-
 export function LemonSelectMultiple({
+    placeholder,
+    options = [],
     value,
-    options,
-    disabled,
     loading,
-    placeholder,
-    labelInValue,
     onChange,
-    onSearch,
-    onFocus,
-    onBlur,
-    filterOption = true,
-    mode = 'single',
-    selectClassName,
+    onInputChange,
+    mode,
+    disabled,
+    disableFiltering = false,
+    allowCustomValues = false,
     ...props
 }: LemonSelectMultipleProps): JSX.Element {
-    const optionsAsList: LemonSelectMultipleOptionItem[] = Array.isArray(options)
-        ? options
-        : Object.entries(options || {}).map(([key, option]) => ({
-              key: key,
-              ...option,
-          }))
-
-    const antOptions = optionsAsList.map((option) => ({
-        key: option.key,
-        value: option.key,
-        label: option.labelComponent || option.label,
-        labelString: option.label,
-    }))
+    const [showPopover, setShowPopover] = useState(false)
+    const [inputValue, setInputValue] = useState('')
+    const popoverFocusRef = useRef<boolean>(false)
+    const inputRef = useRef<HTMLInputElement>(null)
+    const [selectedIndex, setSelectedIndex] = useState(0)
+    const values = value ?? []
+
+    const visibleOptions = useMemo(() => {
+        const res: LemonSelectMultipleOption[] = []
+        const customValues = [...values]
+
+        options.forEach((option) => {
+            // Remove from the custom values list if it's in the options
+
+            if (customValues.includes(option.key)) {
+                customValues.splice(customValues.indexOf(option.key), 1)
+            }
+
+            // Check for filtering
+            if (inputValue && !disableFiltering && !option.label.toLowerCase().includes(inputValue.toLowerCase())) {
+                return
+            }
+
+            res.push(option)
+        })
+
+        // Custom values are always shown before the list
+        if (customValues.length) {
+            customValues.forEach((value) => {
+                res.unshift({ key: value, label: value })
+            })
+        }
+
+        // Finally we show the input value if custom values are allowed and it's not in the list
+        if (allowCustomValues && inputValue && !values.includes(inputValue)) {
+            res.unshift({ key: inputValue, label: inputValue })
+        }
+
+        return res
+    }, [options, inputValue, ...values])
+
+    // Reset the selected index when the visible options change
+    useEffect(() => {
+        setSelectedIndex(0)
+    }, [visibleOptions.length])
+
+    useEffect(() => {
+        onInputChange?.(inputValue)
+    }, [inputValue])
```

Comment:
I'm generally weary of `useEffect` hooks where they can be inserted in the flow of the code because tiny ones like this are so easy to miss & can cause side effects you weren't expecting if you didn't know they were there.

Maybe we could have a wrapper like this instead that gets called everywhere:
```suggestion
    const setInputValue = (value: string): void => {
        setInternalInput(value)
        onInputChange?.(inputValue)
    }
```
(Open to whatever names make sense because of the crossover)

Reply:
Sure